x/evm/keeper/send.go

```diff
@@ -0,0 +1,55 @@
+package keeper
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/sei-protocol/sei-chain/x/evm/types"
+)
+
+func (k *Keeper) EVMToEVMSend(ctx sdk.Context, fromAddress common.Address, toAddress common.Address, amount uint64) error {
+	if err := k.DebitAddress(ctx, fromAddress, amount); err != nil {
+		return err
+	}
+	return k.CreditAddress(ctx, toAddress, amount)
+}
+
+func (k *Keeper) BankToEVMSend(ctx sdk.Context, bankAddress sdk.AccAddress, toAddress common.Address, amount uint64) error {
+	// first transfer tokens to evm module account
+	if err := k.bankKeeper.SendCoinsFromAccountToModule(ctx, bankAddress, types.ModuleName,
+		sdk.NewCoins(sdk.NewCoin(k.GetBaseDenom(ctx), sdk.NewIntFromUint64(amount)))); err != nil {
+		return err
+	}
+	// then credit the recipient address
+	return k.CreditAddress(ctx, toAddress, amount)
+}
+
+func (k *Keeper) EVMToBankSend(ctx sdk.Context, evmAddress common.Address, toAddress sdk.AccAddress, amount uint64) error {
+	// first transfer tokens out of evm module account
+	if err := k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, toAddress,
+		sdk.NewCoins(sdk.NewCoin(k.GetBaseDenom(ctx), sdk.NewIntFromUint64(amount)))); err != nil {
+		return err
+	}
+	// then debit the sender address
+	return k.DebitAddress(ctx, evmAddress, amount)
```

Comment:
do we need to validate that we can debit the amount before doing the send in line 30?

Reply:
evm uses the same cache state as normal tx so any error would cause all state changes in a tx to be rolled back